### PR TITLE
Optional stacktrace generation with libbacktrace

### LINKED ISSUE #1519: Stacktrace in testprecice for triggered PRECICE_ASSERT is not helpful for debugging anymore
**Describe your setup**

Operating system (e.g. Linux distribution and version): _Ubuntu 22.04_
preCICE Version: e3db3c1c7216d8d48564a6ffb2e62b1343bc13a5

**Describe the problem**

If `testprecice` runs into an assertion inside the core library the stacktrace is not useful, because it only showes `testprecice`. I remember that this was different about 2-3 weeks ago, where a hit assertion resulted in a stacktrace that showed me the trace from `testprecice` down to the function that triggered the assertion.

**Step To Reproduce**
1. Introduced an assertion that gets triggered by test `A`. (I added `PRECICE_ASSERT(false, "Assertion triggered in ParallelCouplingScheme::ParallelCouplingScheme")` in the constructor of `ParalleCouplingScheme`)
2. Run `testprecice` for test `A`

For my example:
```
precice/build$ mpirun -np 4 ./testprecice -t CplSchemeTests/ParallelImplicitCouplingSchemeTests/Extrapolation/FirstOrderWithAcceleration
...
testprecice: /home/benjamin/Programming/precice/src/cplscheme/ParallelCouplingScheme.cpp:26: precice::cplscheme::ParallelCouplingScheme::ParallelCouplingScheme(double, int, double, int, const string&, const string&, const string&, precice::m2n::PtrM2N, precice::cplscheme::constants::TimesteppingMethod, precice::cplscheme::BaseCouplingScheme::CouplingMode, int, int): Assertion `false' failed.
ASSERTION FAILED
Location:   precice::cplscheme::ParallelCouplingScheme::ParallelCouplingScheme(double, int, double, int, const string&, const string&, const string&, precice::m2n::PtrM2N, precice::cplscheme::constants::TimesteppingMethod, precice::cplscheme::BaseCouplingScheme::CouplingMode, int, int)
File:       /home/benjamin/Programming/precice/src/cplscheme/ParallelCouplingScheme.cpp:26
Expression: false
Rank:       0
Arguments:  
  0: "Assertion triggered in ParallelCouplingScheme::ParallelCouplingScheme" == Assertion triggered in ParallelCouplingScheme::ParallelCouplingScheme

Stacktrace:
 0# 0x000055D21D0365C0 in ./testprecice
 1# 0x000055D21CD5C7B7 in ./testprecice
 2# 0x000055D21C517783 in ./testprecice
 3# 0x000055D21C5168AD in ./testprecice
 4# 0x000055D21C3A3331 in ./testprecice
 5# 0x00007F5E30C523F2 in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
 6# boost::execution_monitor::catch_signals(boost::function<int ()> const&) in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
 7# boost::execution_monitor::execute(boost::function<int ()> const&) in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
 8# boost::execution_monitor::vexecute(boost::function<void ()> const&) in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
 9# boost::unit_test::unit_test_monitor_t::execute_and_translate(boost::function<void ()> const&, unsigned long) in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
10# 0x00007F5E30C605A9 in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
11# 0x00007F5E30C608BB in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
12# 0x00007F5E30C608BB in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
13# 0x00007F5E30C608BB in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
14# 0x00007F5E30C608BB in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
15# boost::unit_test::framework::run(unsigned long, bool) in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
16# boost::unit_test::unit_test_main(bool (*)(), int, char**) in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
17# 0x000055D21C34E5B5 in ./testprecice
18# __libc_start_main in /lib/x86_64-linux-gnu/libc.so.6
19# 0x000055D21C34E2DE in ./testprecice
```

**Expected behaviour**
The stacktrace should look like
```
....
File:       /home/benjamin/Programming/precice/src/cplscheme/ParallelCouplingScheme.cpp:26
Expression: false
Rank:       0
Arguments:  
  0: "Assertion triggered in ParallelCouplingScheme::ParallelCouplingScheme" == Assertion triggered in ParallelCouplingScheme::ParallelCouplingScheme

Stacktrace:
 0# 0x000055D21D0365C0 in ./ParallelCouplingScheme
 1# 0x000055D21CD5C7B7 in ./SomeThingMore
 2# 0x000055D21C517783 in ./SomeOtherThing
 3# 0x000055D21C5168AD in ./SomeThing
 4# 0x000055D21C3A3331 in ./testprecice
 5# 0x00007F5E30C523F2 in /lib/x86_64-linux-gnu/libboost_unit_test_framework.so.1.71.0
```

**Additional context**

I'm currently suspecting bfb10564b73a5d43215233764fec76bd25ade27b, b8f96f9aefe3916835d23c2f05fc4ecdce8db01c or b8fdde35054d1a287a3a430912ed4a32f103c53b to have introduced this change, but I'm not very deep into CMake so I was not able to fix the problem.

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

Add optional (disabled by default) support for stacktraces with libbacktrace.

To use, enable the CMake option `PRECICE_ENABLE_LIBBACKTRACE`.
This will extract and install libbacktrace from the thirdparty folder and setup boost.stacktrace.
The library will be statically linked against preCICE.

## Motivation and additional information

Since #1498, the default method of stacktrace generation doesn't generate usable stacktraces as the default implementation ignores debug symbols. 

Motivation for adding the library to the thirdparty folder is that this is a developer only tool, which is disabled by default. No distro packages libbacktrace and it has not versioning apart from the commit history.

Closes #1519

## Author's checklist

* [ ] I used the [`pre-commit` hook](https://precice.org/dev-docs-dev-tooling.html#setting-up-pre-commit) to prevent dirty commits and used `pre-commit run --all` to format old commits.
* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [ ] I added a test to cover the proposed changes in our test suite.
* [ ] For breaking changes: I documented the changes in the appropriate [porting guide](https://precice.org/couple-your-code-porting-overview.html).
* [ ] I sticked to C++17 features.
* [ ] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

**Comment by fsimonis** [comment]
Would `PRECICE_ENABLE_LIBBACKTRACE` be a better name compared to the current `PRECICE_ENABLE_BACKTRACE`?

**Comment by BenjaminRodenberg** [comment]
> Would `PRECICE_ENABLE_LIBBACKTRACE` be a better name compared to the current `PRECICE_ENABLE_BACKTRACE`?

Spontaneous thought: Backtrace is what we do. Libbacktrace is just the current implementation. Imagine we switch to a different implementation (using for example `libawesometraces`, which I just made up) we would need to change the name of the option.

**Comment by fsimonis** [comment]
An alternative to using non-existent packages could be to download and compile libbacktrace as an externalproject in the build directory. This way the developer doesn't to handle the entire setup of the pkg-config file etc.

**Comment by fsimonis** [comment]
Just talked to @uekerman. We will try to integrate libbacktrace directly as a third party library

**Comment by fsimonis** [comment]
I added libbacktrace to the thirdparty repo and updated the option name.

**Review by BenjaminRodenberg — CHANGES_REQUESTED** [review]
I merged this PR into #1523 and got the following errors when trying to build:

```
precice/build$ cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DPRECICE_PythonActions=off -DPRECICE_ENABLE_LIBBACKTRACE=ON ..

precice/build$ make -j8
...
-- === ENABLED FEATURES ===
...
 * libbacktrace, Enables libbacktrace for stracktrace generation.

   preCICE assertions are helpful during development, however, the default stacktraces may not be sufficent.
   This feature enables the support for libbacktrace in boost.stacktrace.

   This feature can be enabled/disabled by setting the PRECICE_ENABLE_LIBBACKTRACE CMake option.
...
[ 34%] Building CXX object CMakeFiles/preciceCore.dir/extras/bindings/fortran/src/SolverInterfaceFortran.cpp.o
[ 34%] Built target preciceCore
Scanning dependencies of target precice
Scanning dependencies of target testprecice
make[2]: *** No rule to make target 'thirdparty/libbacktrace/libbacktrace.a', needed by 'libprecice.so.2.5.0'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:278: CMakeFiles/precice.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 35%] Building CXX object CMakeFiles/testprecice.dir/src/testing/main.cpp.o
...
[ 99%] Building CXX object CMakeFiles/testprecice.dir/tests/serial/whitebox/TestConfigurationComsol.cpp.o
make[2]: *** No rule to make target 'thirdparty/libbacktrace/libbacktrace.a', needed by 'testprecice'.  Stop.
make[2]: *** Waiting for unfinished jobs....
[ 99%] Building CXX object CMakeFiles/testprecice.dir/tests/serial/whitebox/TestConfigurationPeano.cpp.o
make[1]: *** [CMakeFiles/Makefile2:386: CMakeFiles/testprecice.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```

After configuring with `-DPRECICE_ENABLE_LIBBACKTRACE=OFF`, building works (but, I guess no stacktrace).

**Comment by fsimonis** [comment]
libbacktrace should be build when running cmake.
You should the find the header and the archive in the build directory in thirdparty/libbacktrace.

Can you check this happens?

**Comment by BenjaminRodenberg** [comment]
> libbacktrace should be build when running cmake. You should the find the header and the archive in the build directory in thirdparty/libbacktrace.
> 
> Can you check this happened? You may need to configure from scratch.

Looks quite empty, but I realized that there also were a few errors:

```
precice$ rm -r build/
precice$ mkdir build
precice$ cd build/
precice/build$ cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DPRECICE_PythonActions=off -DPRECICE_ENABLE_LIBBACKTRACE=ON ..
...
-- Preparing libbacktrace
-- Extracting sources
CMake Error at thirdparty/libbacktrace/CMakeLists.txt:9 (file):
  file does not recognize sub-command ARCHIVE_EXTRACT


-- Running autoconf
CMake Error at thirdparty/libbacktrace/CMakeLists.txt:18 (execute_process):
  execute_process given unknown argument "COMMAND_ERROR_IS_FATAL".


-- Running ./configure
CMake Error at thirdparty/libbacktrace/CMakeLists.txt:18 (execute_process):
  execute_process given unknown argument "COMMAND_ERROR_IS_FATAL".


-- Running make
CMake Error at thirdparty/libbacktrace/CMakeLists.txt:18 (execute_process):
  execute_process given unknown argument "COMMAND_ERROR_IS_FATAL".


-- Copying files
CMake Error at thirdparty/libbacktrace/CMakeLists.txt:26 (file):
  file COPY cannot find
  "/home/benjamin/Programming/precice/build/thirdparty/libbacktrace/libbacktrace-master/backtrace.h":
  No such file or directory.


-- libbacktrace built successfully
...
-- Configuring incomplete, errors occurred!
precice/build$ ls thirdparty/
fmt/          json/         libbacktrace/ 
precice/build$ ls thirdparty/libbacktrace/
CMakeFiles
precice/build$ ls thirdparty/libbacktrace/CMakeFiles/

```

File: CMakeLists.txt [hunk 1]
```diff
@@ -237,6 +246,11 @@ else()
   message(STATUS "Python support disabled")
 endif()
 
+# Option ENABLE_BACKTRACE
```
Comment (BenjaminRodenberg) [inline CMakeLists.txt:249]:
```suggestion
# Option ENABLE_LIBBACKTRACE
```

**Comment by fsimonis** [comment]
Aha we are still stuck with cmake 3.16.3....

I'll come up with a backport and enable this in the CI

File: CMakeLists.txt [hunk 2]
```diff
@@ -63,6 +63,7 @@ option(PRECICE_ALWAYS_VALIDATE_LIBS "Validate libraries even after the validatat
 option(PRECICE_ENABLE_C "Enable the native C bindings" ON)
 option(PRECICE_ENABLE_FORTRAN "Enable the native Fortran bindings" ON)
 option(PRECICE_BUILD_TOOLS "Build the \"precice-tools\" executable" ON)
+option(PRECICE_ENABLE_LIBBACKTRACE "Enable libbacktrace for stacktrace generation." OFF)
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply CMakeLists.txt:66]: Sadly no and this is also quite confusing. There is backtrace (which is found by CMakes FindBacktrace) and there is libbacktrace. Both have different interfaces.
Boost stacktrace uses the libbacktrace interface.

**Comment by fsimonis** [comment]
I backported the CMake to 3.16 and enabled libbacktrace in the linux CI.

**Review by BenjaminRodenberg — APPROVED** [review]
Now everything works nicely :+1: 

Last open point: We should also make sure that this works by enabling it in the CI. Output is useful in the logs and we also get a warning early if this feature (which is probably not used by everybody) breaks.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply CMakeLists.txt:66]: Ok. I think I understand the reason.

**Comment by BenjaminRodenberg** [comment]
@fsimonis Currently there is an error in the CI pipeline: https://github.com/precice/precice/actions/runs/3969218777/jobs/6803414018#step:11:177

As far as I remember, I did not observe these kind of problems. If I can help here, please let me know. Would be great, if we could merge this PR soon, because the stacktrace is superuseful.

**Comment by fsimonis** [comment]
The symlink chain of libbacktrace is incorrect, hence the missing `so.0`.
It should be a chain `libbacktrace.so -> .so.0 -> .so.0.0 -> libbacktrace.so.0.0.0`, but they all link directly to `.so.0.0.0`.

**Comment by BenjaminRodenberg** [comment]
@fsimonis Is this PR ready to merge?

**Pull request merged**